WordPress/gutenberg
# Inserter: Fix 'Browse All' in Quick Inserter

Fixes https://github.com/WordPress/gutenberg/issues/24403.
Related to https://github.com/WordPress/gutenberg/pull/24285.

![insertion-point](https://user-images.githubusercontent.com/612155/97139169-02950100-17ae-11eb-89cc-22e089fcad6e.gif)

Maintain the insertion point in `@wordpress/block-editor` state when moving from the Quick Inserter to the Inserter.

This fixes a bug (https://github.com/WordPress/gutenberg/issues/24403) where the insertion point is wrong after clicking a block appender (e.g. in a Columns block) and selecting _Browse All_.

Care is taken to not regress a previous bug (https://github.com/WordPress/gutenberg/issues/24262) where the insertion point was wrong after clicking an inline insertion button and selecting _Browse All_.

I re-used the existing `insertionPoint` state tree so as to minimise the amount of new actions and selectors needed. This means we have three insertion point actions:

- `setInsertionPoint`: Sets where the inserter will default to inserting blocks.
- `showInsertionPoint`: Sets where the inserter will default to inserting blocks and display this to the user.
- `hideInsertionPoint`: Hides the insertion point from the user.

The insertion point is cleared when block selection changes. This is needed for the Site Editor where the Inserter is persistent.

### Testing

A regression E2E test has been added. This complements the one added in https://github.com/WordPress/gutenberg/pull/24396.

The manual flows that I checked are:

1. Create a new post.
1. Insert a Columns block with a paragraph in the first column.
1. Insert a block at the end of the post.
1. Click the block appender in the empty column.
1. Select _Browse All_.
1. Click on a block. It should be inserted into the empty column.
1. Click on the inline inserter that appears when you hover over the paragraph in the first column.
1. Select _Browse All_.
1. Click on a block. It should be inserted into the first column.

And:

1. Go to the experimental site editor.
1. Hover over a block and click on the inline inserter button that appears.
1. Select _Browse All_.
1. Click on a block. It should be inserted into the location you clicked on.
1. Hover over a block and click on the inline inserter button that appears.
1. Select _Browse All_.
1. Select a different block somewhere else in the page.
1. Click on a block. It should be inserted after the block you selected.
